Squeezebox config flow

rajlaud opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
Configuration of the squeezebox integration through configuration.yaml is now deprecated. Please consider using the frontend to configure the squeezebox integration instead. 


## Proposed change
This moves the squeezebox integration to the configuration flow so that it can be configured from the frontend and loaded without restarting home assistant. Servers will automatically be discovered by the "discovery" integration, or by this integration once it is loaded. If the user adds this integration from frontend, it will attempt to automatically discover the server as well.

Relatedly, the squeezebox integration will now handle individual players coming online or going offline while home assistant is running. Previously, the integration would only detect players that were present during startup.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
Tests have been added for the configuration flow.

- This PR fixes or closes issue: fixes #36843
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13452

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rajlaud committed:
Squeezebox add config flow and player discovery

rajlaud committed:
Fixes to config flow

rajlaud committed:
Unavailable player detection and recovery

rajlaud committed:
Improved error message for auth failure

rajlaud committed:
Testing for squeezebox config flow

rajlaud committed:
Import configuration.yaml

rajlaud committed:
Support for discovery integration

rajlaud committed:
Internal server discovery

rajlaud committed:
Fix bug restoring previously detected squeezebox player

rajlaud committed:
Tests for user and edit steps in config flow

rajlaud committed:
Tests for import config flow

rajlaud committed:
Additional config flow tests and fixes

rajlaud committed:
Linter fixes

rajlaud committed:
Check that players are found before iterating them

rajlaud committed:
Remove noisy logger message

rajlaud commented:
I updated rebased this PR since the pysqueezebox dependency bump (#35956) got merged separately.

rajlaud committed:
Update requirements_all after rebase

bdraco commented on homeassistant/components/squeezebox/config_flow.py:
```diff
@@ -0,0 +1,254 @@
+            async_discover(_discovery_callback)
+        )
+        while not self.discovery_info:
+            await asyncio.sleep(0)
```
`_discovery_callback` should set an event, and we should await the event with a timeout instead.

rajlaud replied:
Got it. Fixed in latest push.

bdraco commented on homeassistant/components/squeezebox/strings.json:
```diff
@@ -0,0 +1,32 @@
+      "edit": {
+        "title": "Edit connection information",
+        "data": {
+          "host": "Host",
```
Please reformat the json and use common keys here

`git grep 'key:common'`

rajlaud replied:
Got it. Added the common keys I could find.

rajlaud replied:
What is the proper way to update translations/en.json? I ran `python3 -m script.translations develop` but it did not update the common keys.

bdraco replied:
I usually do it manually, but this might work:
`python -m script.translations develop --integration squeezebox`

rajlaud replied:
Thanks, I gave it a try but I guess the script doesn't handle these keys. I updated manually.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -126,18 +141,75 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+            if entity and not entity.available:
+                # check if previously unavailable player has connected
+                await player.async_update()
+                entity.available = player.connected
```
Can we avoid having to set `available` from outside the object and have `async_update` check `self._player.connected` instead?

rajlaud replied:
Unfortunately, no. This is a problem with how Logitech Media Server handles disconnected players. For a short period of time (a few minutes), `connected` is false. After that period of time elapses, the server forgets completely about the player, and abruptly disconnects if you attempt to query information about it.

My workaround is stop trying to query disconnected players by having `async_update` skip them. If they are re-discovered, I mark them `connected` so we start updating them again. I check `player.connected` just in case we have rediscovered a player during the limited time it is found by the server but in the disconnected state.

MartinHjelmare replied:
We want to avoid accessing the entity from outside. We can use our dispatch helper to signal the an entity method from outside.

rajlaud replied:
Ok. I think this got merged before your comment, but I can create a new PR to address this and your other comments below.

rajlaud committed:
Use asyncio.Event in discovery task

rajlaud committed:
Use common keys in strings.json

rajlaud committed:
Bump pysqueezebox to v0.2.2 for fixed server discovery using python3.7

rajlaud committed:
Bump pysqueezebox version to v0.2.3

rajlaud commented:
Updated pysqueezebox version - see https://github.com/rajlaud/pysqueezebox/releases/tag/v0.2.3. Fixes https://github.com/home-assistant/core/issues/36843.

rajlaud commented:
Rebased and force-pushed to resolve merge conflict in discovery/__init.py.

bdraco commented on homeassistant/components/squeezebox/config_flow.py:
```diff
@@ -0,0 +1,256 @@
+            DATA_SCHEMA(config)
+            info = await validate_input(self.hass, config)
+            if "uuid" in info:
+                try:
+                    await self.async_set_unique_id(info["uuid"])
+                    # update with info from configuration.yaml
+                    self._abort_if_unique_id_configured(info)
+                except data_entry_flow.AbortFlow as error:
+                    return self.async_abort(reason=error.reason)
```
```suggestion
                await self.async_set_unique_id(info["uuid"])
                # update with info from configuration.yaml
                self._abort_if_unique_id_configured(info)
```

If you don't trap the exception it will abort on its own

bdraco commented on homeassistant/components/squeezebox/config_flow.py:
```diff
@@ -0,0 +1,256 @@
+        """Edit a discovered or manually inputted server."""
+        errors = {}
+        if user_input:
+            try:
+                info = await validate_input(self.hass, user_input)
+                if "uuid" in info:
+                    await self.async_set_unique_id(info["uuid"])
+                    self._abort_if_unique_id_configured()
+                return self.async_create_entry(title=info.get("ip"), data=user_input)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                errors["base"] = "unknown"
```
This code is similar to `async_step_import`.

Can you break it out into a new function, return the error as `error_string`, and then do something like use `return self.async_abort(reason=error_string)` in `async_step_import` and `errors["base"] = error_string` in `async_step_edit`?

rajlaud replied:
Makes sense. I'll make `validate_input` a method of `SqueezeboxConfigFlow` and then it can perform all of these tasks.

rajlaud committed:
Don't trap AbortFlow exception

Co-authored-by: J. Nick Koston <nick@koston.org>

rajlaud committed:
Refactor validate_input

bdraco commented on homeassistant/components/squeezebox/config_flow.py:
```diff
@@ -0,0 +1,230 @@
+        discovery_task.cancel()  # stop searching as soon as we find server
+
+        # update with suggested values from discovery
+        self.data_schema = vol.Schema(
```
There is a lot of similar code to build the schema.  You can pass discovery_info/user_input to a function to build it in pieces to avoid repeating yourself.

Example: https://github.com/home-assistant/core/pull/36827/commits/fc352e93b020547a6afd8459ac65ff4fdd0bdaab

rajlaud replied:
Ok, implemented a `base_schema` function that builds the schema based on `discovery_info` or `user_input`.

bdraco replied:
It looks like there is still quite a bit of schema duplication

rajlaud committed:
Update squeezebox tests

rajlaud committed:
Build data flow schema using function

rajlaud committed:
Fix linter error

rajlaud committed:
Updated en.json

bdraco commented:
`.coveragerc` needs to be updated to so `config_flow.py` is included.

`pytest --cov=homeassistant/components/squeezebox/ --cov-report term-missing -- tests/components/squeezebox/test_*.py`

The test is taking a while so there is probably some unmocked I/O

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,52 +71,61 @@
+
+    if config:
+        conf = {
+            CONF_HOST: config.get(CONF_HOST),
```
```suggestion
            CONF_HOST: config[CONF_HOST],
```

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,52 +71,61 @@
+    if config:
+        conf = {
+            CONF_HOST: config.get(CONF_HOST),
+            CONF_PORT: config.get(CONF_PORT) if config.get(CONF_PORT) else DEFAULT_PORT,
```
```suggestion
            CONF_PORT: config.get(CONF_PORT, DEFAULT_PORT),
```

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,52 +71,61 @@
 
     username = config.get(CONF_USERNAME)
     password = config.get(CONF_PASSWORD)
+    host = config.get(CONF_HOST)
```
```suggestion
    host = config[CONF_HOST]
```

Host is required right?

rajlaud replied:
Yes.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,52 +71,61 @@
-    else:
-        host = config.get(CONF_HOST)
-        port = config.get(CONF_PORT)
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
```
```suggestion
    hass.data.setdefault(DOMAIN, {})
```

rajlaud committed:
Update homeassistant/components/squeezebox/media_player.py

Co-authored-by: J. Nick Koston <nick@koston.org>

rajlaud committed:
Update homeassistant/components/squeezebox/media_player.py

Co-authored-by: J. Nick Koston <nick@koston.org>

rajlaud committed:
Update homeassistant/components/squeezebox/media_player.py

Co-authored-by: J. Nick Koston <nick@koston.org>

rajlaud committed:
Update homeassistant/components/squeezebox/media_player.py

Co-authored-by: J. Nick Koston <nick@koston.org>

rajlaud committed:
Update .coveragerc for squeezebox config flow test

rajlaud committed:
Mock TIMEOUT for faster testing

rajlaud committed:
More schema de-duplication and testing improvements

rajlaud commented:
Fixed the delay - the I/O was mocked but timeout was left at 5 seconds. Added additional tests.

Also found more duplicated schema as I was going through the tests. Should be all cleaned up now.

bdraco commented on homeassistant/components/squeezebox/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+            await asyncio.wait_for(self._discover(), timeout=TIMEOUT)
+            return await self.async_step_edit()
+        except asyncio.TimeoutError:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=vol.Schema({vol.Optional(CONF_HOST): str}),
+                errors={"base": "no_server_found"},
+            )
```
```suggestion
            errors["base"] = "no_server_found"
```

Can we fall though here since it will show the form?

bdraco commented on homeassistant/components/squeezebox/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+
+    async def async_step_import(self, config):
+        """Import a config flow from configuration."""
+        _base_schema()(config)
```
It looks like this is validating the schema here.  We should leave the platform schema for now since it will do the validation of the yaml which means we will always have valid data to import.

rajlaud replied:
Ok, will remove.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,52 +71,60 @@
 async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Set up the squeezebox platform."""
+    """Set up squeezebox platform from platform entry in configuration.yaml (deprecated)."""
+    _LOGGER.warning("Loading Squeezebox via platform config is deprecated.")
```
Please use cv.deprecated in the platform schema instead.

rajlaud replied:
I removed the logger message, but I'm confused about the best way to use `cv.deprecated` here. The goal is to direct people to the config flow, not warn that certain options are not deprecated. I plan to support `configuration.yaml` for the foreseeable future, so maybe no warning is needed.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,52 +71,60 @@
     username = config.get(CONF_USERNAME)
     password = config.get(CONF_PASSWORD)
+    host = config[CONF_HOST]
+    port = config.get(CONF_PORT)
```
```suggestion
    port = config[CONF_PORT]
```

Port is never going to be unset right?

rajlaud replied:
Correct.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -126,18 +140,75 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+                    )
+                )
+
+        if DOMAIN not in hass.data:
+            hass.data[DOMAIN] = {}
```
```suggestion
```

Already done in async_setup_entry

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -126,18 +140,75 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+    _LOGGER.debug("Adding player discovery job for LMS server: %s", ipaddr)
+    hass.async_add_job(_discovery)
+
+    @callback
```
Since this is `async` it seem like it doesn't need to be decorated with `@callback`?

rajlaud replied:
Agreed. Removing decorators.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -126,18 +140,75 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+    async def start_server_discovery():
+        """Start a server discovery task."""
+
+        @callback
```
Since this is `async` it seem like it doesn't need to be decorated with `@callback`?

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -126,18 +140,75 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+    hass.async_add_job(_discovery)
+
+    @callback
+    async def start_server_discovery():
```
Does this need to be nested in `async_setup_entry`?

rajlaud replied:
No, moving this to a standalone function.

rajlaud replied:
Also, I moved the start job to the integration setup; this makes more sense than waiting for at least one valid config entry to start searching.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -126,18 +140,75 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
-    media_players = []
-    for player in players:
-        media_players.append(SqueezeBoxDevice(player))
+    async def _discovery(now=None):
```
Does this need to be nested in `async_setup_entry`?

rajlaud replied:
Yes. It is much easier for the player discovery functions to be nested - than we can use `async_call_later` without parameters. This thread is specific to the configuration entry (the individual Logitech Media Server).

rajlaud committed:
Apply suggestions from code review

Co-authored-by: J. Nick Koston <nick@koston.org>

rajlaud committed:
Testing and config flow improvements

rajlaud committed:
Remove unused exceptions

rajlaud committed:
Remove deprecated logger message

rajlaud commented:
Despite setting a `unique_id`, the "ignore" button is missing from the frontend. I can't seem to figure out why. Any thoughts?

rajlaud commented:
I see that "discovery" is not in the list of sources that allow the "ignore" button to appear in the frontend. See https://github.com/home-assistant/frontend/blob/dev/src/data/config_flow.ts

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,79 +67,118 @@
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up an LMS Server from a config entry."""
+    config = config_entry.data
+    _LOGGER.debug("Reached async_setup_entry, config=%s", config)
```
Security: We shouldn't log the password here.

rajlaud replied:
Got it. Will just log host.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,79 +67,118 @@
+    """Start a server discovery task."""
+
+    def _discovered_server(server):
+        hass.loop.create_task(
```
```suggestion
        asyncio.create_task(
```

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,79 +67,118 @@
+}
+
+
+async def start_server_discovery(hass):
```
It looks like this function is only called in `__init__.py` so it might be better off there.

rajlaud replied:
Ok, moved it.

bdraco commented on homeassistant/components/squeezebox/__init__.py:
```diff
@@ -1 +1,28 @@
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, MP_DOMAIN)
+    )
+    return True
```
`async_unload_entry` is missing here.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,79 +67,118 @@
     | SUPPORT_CLEAR_PLAYLIST
 )
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
We should leave the platform schema and add `cv.deprecated` so they still get validation of the yaml during the transition period.

rajlaud replied:
Ah, got it. And now I see where cv.deprecated goes. Thanks.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,79 +67,118 @@
+        hass.helpers.event.async_call_later(DISCOVERY_INTERVAL, _discovery)
+
+    _LOGGER.debug("Adding player discovery job for LMS server: %s", host)
+    hass.async_add_job(_discovery)
```
`async_add_job` is legacy and shouldn't be used in new code.

rajlaud replied:
Ok. Since _discovery is a coroutine, I used asyncio.create_task instead.

rajlaud committed:
Update homeassistant/components/squeezebox/media_player.py

Co-authored-by: J. Nick Koston <nick@koston.org>

rajlaud committed:
Implement suggestions from code review

rajlaud committed:
Add async_unload_entry

rajlaud commented:
Test failures are unrelated to this integration/PR - not sure how to get the tests to re-run.

bdraco commented:
/AzurePipelines Run

bdraco commented:
If it doesn't pass, you may need to rebase
https://developers.home-assistant.io/docs/development_catching_up/

rajlaud commented:
Thanks. I think the remaining failure was just a timeout, but I rebased anyway.

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,287 @@
+UNKNOWN_ERROR = "1234"
+
+
+class MockEntry:
```
There is an existing fixture for these
`from tests.common import MockConfigEntry`

rajlaud replied:
Nice. I switched it to this.

bdraco commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,79 +68,108 @@
+    """Set up squeezebox platform from platform entry in configuration.yaml (deprecated)."""
+
+    if config:
+        conf = {
+            CONF_HOST: config[CONF_HOST],
+            CONF_PORT: config.get(CONF_PORT, DEFAULT_PORT),
+        }
+        for attr in [CONF_USERNAME, CONF_PASSWORD]:
+            if attr in config:
+                conf.update(attr, config.get(attr))
```
Is this needed anymore now that the schema is back?

rajlaud replied:
You're right, it's not needed anymore. We can pass the platform schema directly.

rajlaud committed:
Use MockConfigEntry in squeezebox tests

rajlaud committed:
Remove unnecessary config schema

bdraco commented on homeassistant/components/squeezebox/__init__.py:
```diff
@@ -1 +1,64 @@
+
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    return await hass.config_entries.async_forward_entry_unload(entry, MP_DOMAIN)
```
Is discovery still going to be running if all entries are deleted?

rajlaud replied:
Yes. The server discovery will run if the integration is loaded. I think this behavior makes sense. For example, if the server `host`, `username`, or `password` changed, the user would delete the old entry and add a new one.

bdraco replied:
Ideally we stop discovery if there are no more config entries loaded, and then restart it the next time a config entry it loaded.

This allows users to migrate from a squeezebox to some other device without restarting and having the overhead of discovery running for a device they no longer have.

rajlaud replied:
Ok. The discovery task will now end on unload of the last config entry.

bdraco replied:
Thanks for fixing that.  We probably need to start it again if it’s not running when the first config entry is loaded instead of in async setup?

rajlaud replied:
You're right - we need to make sure it is running when we load a config entry. One nice thing about having it start in `async_setup` is that you can load the integration manually in `configuration.yaml` with `squeezebox:` and let it auto-discover squeezeboxes (which is a good substitute for using the deprecated `discovery` integration if `squeezebox` is the only thing left in `discovery` you care about).

I also realized the player discovery task was not cleaning up gracefully, so I fixed that too.

rajlaud committed:
Stop server discovery task when last config entry unloaded

rajlaud committed:
Improvements to async_unload_entry

bdraco commented:
I think there is one more use case.  

1. Setup squeezebox config entry.
2. Delete config entry
3. Add another config entry.

I'm not sure if `async_setup` will be run again so discover might not ever startup again in this instance.

Everything else is looking good at this point, I'll run though it one more time once the above is addressed or confirmed to not be an issue.

rajlaud committed:
Fix bug in _discovery arguments

rajlaud commented:
I think that use case is addressed by `start_server_discovery` in `async_setup_entry` as well as `async_setup`. `start_server_discovery` checks if server discovery is already running so we can safely call it in both places. I added that in 60b0214.

However, I realized I did introduce a bug by deleting the unused variable `now` from the definition of the player `_discovery` function in f9877dd. I forgot that `async_call_later` ends up passing this argument, so we have to be prepared to accept it. I just fixed that.

Thanks for all your help on this. I've definitely learned a lot from your code review.

bdraco commented on homeassistant/components/squeezebox/__init__.py:
```diff
@@ -1 +1,76 @@
+        )
+
+    hass.data.setdefault(DOMAIN, {})
+    if DISCOVERY_TASK not in hass.data[DOMAIN]:
```
There is a `TOCTOU` race condition here since there isn't an `asyncio.Lock` when checking to see if the discovery task is in `hass.data[DOMAIN]`

bdraco replied:
Disregard, I don't think it can actually ever happen.

rajlaud replied:
Ah. If I instantiate an `asyncio.Lock` and then prefix the `if DISCOVERY_TASK` block with `async with lock:` that should remove the race condition, right? It's not hypothetical since `start_server_discovery` is getting called from both `async_setup` and `async_setup_entry`.

bdraco commented on homeassistant/components/squeezebox/__init__.py:
```diff
@@ -1 +1,76 @@
+    hass.data[DOMAIN][PLAYER_DISCOVERY_UNSUBS][entry.unique_id]()
+
+    # Stop server discovery task if this is the last config entry.
+    current_entries = hass.config_entries.async_entries(DOMAIN)
```
The same `TOCTOU` race condition mentioned above exists here as well.

bdraco replied:
Disregard, I don't think it can actually ever happen.

rajlaud committed:
Do not await server discovery in async_setup_entry

rajlaud commented:
Ok, thanks. But now I know about `asyncio.Lock()`!

On Sun, Jun 21, 2020 at 11:48 AM J. Nick Koston <notifications@github.com>
wrote:

> *@bdraco* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/squeezebox/__init__.py
> <https://github.com/home-assistant/core/pull/35669#discussion_r443237033>:
>
> > +
> +    def _discovered_server(server):
> +        asyncio.create_task(
> +            hass.config_entries.flow.async_init(
> +                DOMAIN,
> +                context={"source": SOURCE_DISCOVERY},
> +                data={
> +                    CONF_HOST: server.host,
> +                    CONF_PORT: int(server.port),
> +                    "uuid": server.uuid,
> +                },
> +            )
> +        )
> +
> +    hass.data.setdefault(DOMAIN, {})
> +    if DISCOVERY_TASK not in hass.data[DOMAIN]:
>
> Disregard, I don't think it can actually ever happen.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/35669#discussion_r443237033>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMCNERFWFVBXUUYUS3654X3RXY2UHANCNFSM4NBXONEA>
> .
>

rajlaud committed:
Do not await start server discovery in async_setup

rajlaud commented:
Changed `await` to `asyncio.create_task` in both places `start_server_discovery` is called so that the server discovery task does not block the integration from finishing configuration.

rajlaud committed:
Do not start server discovery from async_setup_entry until homeassistant running

rajlaud committed:
Re-detect players when server removed and re-added without restart

rajlaud commented:
One more use case: deleting a config entry and re-adding the same server. I needed to remove the config entry's players from the list of known players so that they would be re-detected without a restart.

bdraco commented on homeassistant/components/squeezebox/__init__.py:
```diff
@@ -64,7 +64,17 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Unload a config entry."""
     # Stop player discovery task for this config entry.
-    hass.data[DOMAIN][PLAYER_DISCOVERY_UNSUBS][entry.unique_id]()
+    hass.data[DOMAIN][entry.unique_id][PLAYER_DISCOVERY_UNSUB]()
```
Usually we store these by `entry.entry_id` instead of `entry.unique_id`

rajlaud replied:
Got it. Thanks.

rajlaud committed:
Use entry.entry_id instead of unique_id

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+
+async def test_import_existing(hass):
+    """Test handling of configuration import of existing server."""
+    with patch(
```
We should use MockConfigEntry here instead

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+async def test_import_bad_auth(hass):
+    """Test handling of configuration import with bad authentication."""
+    with patch(
+        "homeassistant.components.squeezebox.config_flow.SqueezeboxConfigFlow._validate_input",
```
We shouldn't patch homeassistant code, we should instead patch `pysqueezebox`

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+async def test_import_bad_host(hass):
+    """Test handling of configuration imported with bad host."""
+    with patch(
+        "homeassistant.components.squeezebox.config_flow.SqueezeboxConfigFlow._validate_input",
```
We shouldn't patch homeassistant code, we should instead patch `pysqueezebox`

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+async def test_import(hass):
+    """Test handling of configuration imported."""
+    with patch(
+        "homeassistant.components.squeezebox.config_flow.SqueezeboxConfigFlow._validate_input",
```
We shouldn't patch homeassistant code, we should instead patch `pysqueezebox`

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+        "homeassistant.components.squeezebox.config_flow.SqueezeboxConfigFlow._validate_input",
+        return_value=None,
+    ), patch(
+        "homeassistant.components.squeezebox.async_setup", return_value=True
```
Ok to patch `async_setup` and `async_setup_entry` as there isn't a good way to block setup otherwise

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+async def test_discovery_no_uuid(hass):
+    """Test handling of discovered server with unavailable uuid."""
+    with patch(
+        "homeassistant.components.squeezebox.config_flow.SqueezeboxConfigFlow._validate_input",
```
We shouldn't patch homeassistant code, we should instead patch `pysqueezebox`

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+async def test_discovery(hass):
+    """Test handling of discovered server."""
+    with patch(
+        "homeassistant.components.squeezebox.config_flow.SqueezeboxConfigFlow._validate_input",
```
We shouldn't patch homeassistant code, we should instead patch `pysqueezebox`

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+    with patch(
+        "homeassistant.components.squeezebox.config_flow.async_discover", mock_discover,
+    ), patch(
+        "homeassistant.components.squeezebox.config_flow.SqueezeboxConfigFlow._async_current_entries",
```
create a `MockConfigEntry` and then add the entry instead of mocking here.

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+async def test_user_form_timeout(hass):
+    """Test we handle server search timeout."""
+    with patch(
+        "homeassistant.components.squeezebox.config_flow.async_discover",
```
We shouldn't patch homeassistant code, we should instead patch `pysqueezebox`

rajlaud replied:
Got it. This is actually `pysqueezebox` code that I patched after import rather than as `pysqueezebox.async_discover`. Fixing to patch `pysqueezebox` directly.

bdraco replied:
Its fine as-is since the code being patched is `pysqueezebox`

For other reviewers: `from pysqueezebox import Server, async_discover`

rajlaud replied:
Actually, I think because I imported it as `from pysqueezebox import async_discover` I have to patch it as `pysqueezebox.components.squeezebox.config_flow.async_discover`

bdraco commented on tests/components/squeezebox/test_config_flow.py:
```diff
@@ -0,0 +1,283 @@
+async def test_user_form(hass):
+    """Test user-initiated flow, including discovery and the edit step."""
+    with patch(
+        "homeassistant.components.squeezebox.config_flow.SqueezeboxConfigFlow._validate_input",
```
We shouldn't patch homeassistant code, we should instead patch `pysqueezebox`

bdraco reviewed (changes requested):
Sorry I didn't see the mocking was altering the config flow code earlier or I would have pointed it out sooner.

rajlaud commented:
No problem, I'll rework the tests.

rajlaud committed:
Update unittests to avoid patching homeassistant code

MartinHjelmare commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -203,24 +249,38 @@ def unique_id(self):
         """Return the state of the device."""
-        if self._player.power is not None and not self._player.power:
+        if not self.available:
+            return STATE_UNAVAILABLE
```
We don't need to handle state unavailable. It's handled by the base entity class.

MartinHjelmare commented on homeassistant/components/squeezebox/media_player.py:
```diff
@@ -65,79 +76,108 @@
-    except OSError as error:
-        _LOGGER.error("Could not communicate with %s:%d: %s", host, port, error)
-        raise PlatformNotReady from error
+    known_players = hass.data[DOMAIN].get(KNOWN_PLAYERS)
```
We can use `dict.setdefault` here.

rajlaud commented on homeassistant/components/squeezebox/__init__.py:
```diff
@@ -1 +1,86 @@
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Logitech Squeezebox component."""
+    if hass.is_running:
+        asyncio.create_task(start_server_discovery(hass))
```
So that instead of loading the `discovery:` integration, you can load `squeezebox:` in `configuration.yaml` and auto-discover servers. Since `discovery` is deprecated and most other things can be discovered through `ssdp` and `zeroconf`, I thought this was useful.

I run it when setting up a config entry too, so that you automatically begin discovering squeezebox servers once you first load the integration.

`start_server_discovery` checks if server discovery is already running so we won't end up with multiple copies of this running.

MartinHjelmare replied:
We're deprecating the platform section in configuration.yaml here so we shouldn't add new features that depend on some other part of configuration.yaml.

rajlaud replied:
Ok, then I will run the internal discovery for servers only if there is a current config entry. This and other suggested changes in PR #37030

Hypfer commented:
The Squeezebox integration will now handle individual players coming online or going offline while Home Assistant is running. Previously, the integration would only detect players that were present during startup.

🙏

Thank you!
